compiler/fir/semantics/src/org/jetbrains/kotlin/fir/resolve/dfa/cfg/ControlFlowGraph.kt

```diff
@@ -114,32 +117,39 @@ enum class EdgeKind(
     DeadBackward(usedInDfa = false, usedInDeadDfa = false, usedInCfa = true, isBack = true, isDead = true)
 }
 
+private val CFGNode<*>.previousNodeCount
+    get() = previousNodes.count { it.owner == owner && !edgeFrom(it).kind.isBack }
+
 private fun ControlFlowGraph.orderNodes(): List<CFGNode<*>> {
-    val visitedNodes = linkedSetOf<CFGNode<*>>()
-    /*
-     * [delayedNodes] is needed to accomplish next order contract:
-     *   for each node all previous node lays before it
-     */
-    val stack = ArrayDeque<CFGNode<*>>()
-    stack.addFirst(enterNode)
-    while (stack.isNotEmpty()) {
-        val node = stack.removeFirst()
-        val previousNodes = node.previousNodes
-        if (previousNodes.any { it !in visitedNodes && it.owner == this && !node.edgeFrom(it).kind.isBack }) {
-            stack.addLast(node)
-            continue
-        }
-        if (!visitedNodes.add(node)) continue
-
-        // NOTE: this intentionally does not walk through subgraphs. If the only path from A to B
-        //  is through a subgraph, add a dead edge between them to enforce ordering.
-        for (followingNode in node.followingNodes) {
-            if (followingNode.owner == this) {
-                if (followingNode !in visitedNodes) {
-                    stack.addFirst(followingNode)
+    // NOTE: this produces a BFS order. If desired, a DFS order can be created instead by using a linked list,
+    // iterating over `followingNodes` in reverse order, and inserting new nodes at the current iteration point.
```

Comment:
A few checkers use `ControlFlowGraph.traverse` in a suspicious way by pushing/popping scopes when visiting certain nodes. This is not correct: the only invariant topological order is guaranteed to provide is that a node will be ordered after all its predecessors. Those checkers may think that in `if (a) { b } else { c }`, `b` and `c` are somehow in both blocks simultaneously.

I don't know if this commit breaks anything in these checkers. They were probably broken already. The amount of tailrec-related warnings I see in bootstrap builds is suspicious. Checkers that rely on the FIR tree structure should not use the CFG.